### PR TITLE
Eager tempfile deletion without atexit part1

### LINKED ISSUE #4446: ctor_evaller.py leaks ctorEval.js files in the temp directory.
STR:

a) Run `em++ tests\hello_libcxx.cpp -o a.html -Oz` and see the temp directory, which now has 4 files of form `tmpehbajx.ctorEval.js`, or

b) run `EM_TESTRUNNER_DETECT_TEMPFILE_LEAKS=1 python tests/runner.py other.test_warn_toomany_vars`:

	C:\code\emsdk\emscripten\incoming>python tests/runner.py other.test_warn_toomany_vars
	WARNING:root:use EM_ALL_ENGINES=1 in the env to run against all JS engines, which is slower but provides more coverage
	Test suites:
	['test_other']
	Running test_other: (1 tests)
	test_warn_toomany_vars (test_other.other) ... C:\code\emsdk\emscripten\incoming\tests\hello_libcxx.cpp z
	ERROR: After running test, there are 4 new temporary files/directories left behind:
	leaked file: C:\code\emsdk\emsdk_tmp\tmpkc0yi3.ctorEval.js
	leaked file: C:\code\emsdk\emsdk_tmp\tmpxmiklj.ctorEval.js
	leaked file: C:\code\emsdk\emsdk_tmp\tmpswe3yx.ctorEval.js
	leaked file: C:\code\emsdk\emsdk_tmp\tmpagvovm.ctorEval.js
	ERROR

Not sure if this is Windows-specific or if it occurs on other OSes as well.

(running the default and other suites with `EM_TESTRUNNER_DETECT_TEMPFILE_LEAKS=1` seems to find other leaks as well, although note that this option is not compatible with the `parallel_test_core.py` script)

### PR DESCRIPTION
**Pull request by juj**
Makes a number of call sites that use temp files to explicitly manager their use of the files, and immediately clean them up after done. Closes #4446.

This is part 1 of the work, part 2 is less critical but something that I'd like to do in the future, which is to deprecate the `temp_files.note()` altogether and remove relying on Python atexit behavior for the cleanup. There are still a few places where this is done after this PR, but none which currently interact with `multiprocessing.Pool` I think, which is why the `other` suite will successfully run without leaks after this.

Also fixes few real temp file leaks that existed in the test suite and in duplicate_function_eliminator.py.

### CODE REVIEW

File: tests/test_other.py [hunk 1]
```diff
@@ -6387,12 +6418,12 @@ def test_function_eliminator_double_parsed_correctly(self):
       tools.duplicate_function_eliminator.run(output_file)
 
       # Run last opts
-      tools.shared.safe_copy(tools.js_optimizer.run(output_file, ['last', 'asm']), output_file)
-      output_file_contents = self.get_file_contents(output_file)
+#      tools.shared.safe_copy(tools.js_optimizer.run(output_file, ['last', 'asm']), output_file)
```
Comment (kripken) [inline tests/test_other.py]: why are these lines removed?

File: tests/test_other.py [hunk 2]
```diff
@@ -6387,12 +6418,12 @@ def test_function_eliminator_double_parsed_correctly(self):
...
+#      output_file_contents = self.get_file_contents(output_file)
 
       # Compare
-      expected_file_contents = self.get_file_contents(path_from_root('tests', 'optimizer', 'test-function-eliminator-double-parsed-correctly-output.js'))
-      self.assertIdentical(output_file_contents, expected_file_contents)
+#      expected_file_contents = self.get_file_contents(path_from_root('tests', 'optimizer', 'test-function-eliminator-double-parsed-correctly-output.js'))
```
Comment (kripken) [inline tests/test_other.py]: these too?

**Comment by kripken** [comment]
lgtm except for those two comments.

File: tests/test_other.py (hunk 1 above)
Comment (juj) [inline tests/test_other.py]: Err, amazing catch(!) :+1:  These were accidentally left over from debugging which call was doing the leaking, and lost sight of this swallowed by the change in indentation. Fixed
